core/services/synchronization/web_socket_client.go

```diff
@@ -127,27 +154,40 @@ const (
 // lexical confinement of done chan allows multiple connectAndWritePump routines
 // to clean up independent of itself by reducing shared state. i.e. a passed done, not w.done.
 func (w *websocketClient) connectAndWritePump(parentCtx context.Context, wg *sync.WaitGroup) {
-	wg.Done()
+	doneWaiting := false
 	logger.Infow("Connecting to explorer", "url", w.url)
 
 	for {
 		select {
 		case <-parentCtx.Done():
+			if !doneWaiting {
+				wg.Done()
+			}
 			return
 		case <-time.After(w.sleeper.After()):
 			connectionCtx, cancel := context.WithCancel(parentCtx)
 			defer cancel()
 
 			if err := w.connect(connectionCtx); err != nil {
+				w.status = ConnectionStatusError
+				if !doneWaiting {
+					wg.Done()
+				}
 				logger.Warn("Failed to connect to explorer (", w.url.String(), "): ", err)
 				break
 			}
 
+			w.status = ConnectionStatusConnected
+			if !doneWaiting {
+				wg.Done()
+			}
 			logger.Info("Connected to explorer at ", w.url.String())
 			w.sleeper.Reset()
 			go w.readPump(cancel)
 			w.writePump(connectionCtx)
 		}
+
+		doneWaiting = true
```

Comment:
The use of the doneWaiting in the loop is a bit difficult for me to reason about

Reply:
`wg.Done` panics when it gets to a negative counter. So `doneWaiting` is essentially a guard to prevent that.